### PR TITLE
Graph Convolutional Networks: Model Relations In Data

### PR DESCRIPTION
**Pull request by VictorBebnev**

### CODE REVIEW

**Comment by VictorBebnev** [comment]
@sergregory please review these changes.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/README.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Training
+For training run [jupyter notebook](graph_convolutional_networks_model_relations_in_data.ipynb)
+
+_Note: we use dataset from_ [_this post_](../PyTorch-Multi-Label-Image-Classification-Image-Tagging/README.md)
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/README.md:4]: @VictorBebnev, do we need to follow any steps from that post in order to make it work? I'd suggest to write it explicitly. 
> Note: we use the dataset from this post, so please refer to it for a detailed description and instructions.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 2]
```diff
@@ -0,0 +1,692 @@
...
+    "\n",
+    "    file_id = '0B7IzDz-4yH_HMFdiSE44R1lselE'\n",
+    "    path_to_tar_file = str(time.time()) + '.tar.gz'\n",
+    "    download_file_from_google_drive(file_id, path_to_tar_file)\n",
+    "    print('Extraction')\n",
+    "    with tarfile.open(path_to_tar_file) as tar_ref:\n",
+    "        tar_ref.extractall(os.path.dirname(img_folder))\n",
+    "    os.remove(path_to_tar_file)\n",
+    "# Also, copy our pre-processed annotations to the dataset folder. \n",
+    "# Note: you can find script for generating such annotations in attachments\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:106]: There is no such script, it was in the previous post.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 3]
```diff
@@ -0,0 +1,692 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# Download Glove model trained on wikipedia.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:119]: Fix all mentions of [GloVe](https://nlp.stanford.edu/projects/glove/).

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 4]
```diff
@@ -0,0 +1,692 @@
...
+    "\n",
+    "    x = []\n",
+    "    y = []\n",
+    "    for value in new_values:\n",
+    "        x.append(value[0])\n",
+    "        y.append(value[1])\n",
+    "        \n",
+    "    plt.figure(figsize=(13, 13)) \n",
+    "    for i in range(len(x)):\n",
+    "        plt.scatter(x[i],y[i])\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:204]: formatting

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 5]
```diff
@@ -0,0 +1,692 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now we can draw t-SNE visualisation.\n",
+    "tsne_plot(dict(zip(small_labels, vectorised_labels)))"
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:222]: You form a dictionary here but then in the function you return back to two lists. Remove the unnecessary conversion.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 6]
```diff
@@ -0,0 +1,692 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# To proceed with the training we first need to compute adjacency matrix.\n",
+    "adj_matrix_path = 'adjacency_matrix.json'\n",
+    "# Count all labels.\n",
+    "nums = np.sum(np.array(dataset_train.annos), axis=0)\n",
+    "adj = np.zeros((len(small_labels), len(small_labels)), dtype=int)\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:325]: Make len(small_labels) a variable?

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 7]
```diff
@@ -0,0 +1,692 @@
...
+    "            return output + self.bias\n",
+    "        else:\n",
+    "            return output\n",
+    "\n",
+    "    def __repr__(self):\n",
+    "        return self.__class__.__name__ + ' (' \\\n",
+    "               + str(self.in_features) + ' -> ' \\\n",
+    "               + str(self.out_features) + ')'\n",
+    "\n",
+    "# Create adjacency matrix from probabilities.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:384]: What probabilities? We only computed number of pairs of labels so far.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+    "    nums = np.array(adj_data['nums']).astype(np.float32)\n",
+    "    nums = nums[:, np.newaxis]\n",
+    "    adj = adj / nums\n",
+    "    adj[adj < t] = 0\n",
+    "    adj[adj >= t] = 1\n",
+    "    adj = adj * p / (adj.sum(0, keepdims=True) + 1e-6)  \n",
+    "    adj = adj + np.identity(num_classes, np.int)\n",
+    "    return adj\n",
+    "\n",
+    "# Apply adjacency matrix renormalisation.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:396]: re-normalization

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 9]
```diff
@@ -0,0 +1,692 @@
...
+    "        self.sigm = nn.Sigmoid()\n",
+    "\n",
+    "        self.features = models.resnext50_32x4d(pretrained=True)\n",
+    "        self.features.fc = nn.Identity()\n",
+    "        self.num_classes = n_classes\n",
+    "\n",
+    "        self.gc1 = GraphConvolution(in_channel, 1024)\n",
+    "        self.gc2 = GraphConvolution(1024, 2048)\n",
+    "        self.relu = nn.LeakyReLU(0.2)\n",
+    "        # Load probability data for adjacency matrix\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:416]: Again, not probabilities but maybe statistics?

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 10]
```diff
@@ -0,0 +1,692 @@
...
+    "])\n",
+    "\n",
+    "# Train preprocessing.\n",
+    "train_transform = transforms.Compose([\n",
+    "    transforms.Resize((256, 256)),\n",
+    "    transforms.RandomHorizontalFlip(),\n",
+    "    transforms.ColorJitter(),\n",
+    "    transforms.RandomAffine(degrees=20, translate=(0.2, 0.2), scale=(0.5, 1.5),\n",
+    "                            shear=None, resample=False, \n",
+    "                            fillcolor=tuple(np.array(np.array(mean)*255).astype(int).tolist())),\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:530]: formatting?

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 11]
```diff
@@ -0,0 +1,692 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# Run inference on the test data.\n",
+    "model.eval()\n",
+    "for sample_id in [1,2,3,4,6]:\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:653]: formatting

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 12]
```diff
@@ -0,0 +1,692 @@
...
+   "outputs": [],
+   "source": [
+    "# Initialize the training parameters.\n",
+    "num_workers = 8 # Number of CPU processes for data preprocessing\n",
+    "lr = 5e-6 # Learning rate\n",
+    "batch_size = 32\n",
+    "save_freq = 1 # Save checkpoint frequency (epochs)\n",
+    "test_freq = 200 # Test model frequency (iterations)\n",
+    "max_epoch_number = 50 # Number of epochs for training \n",
+    "# Note: on the small subset of data overfitting happens after 30-35 epochs.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:478]: Why we use max_epoch_number=50 if it overfits after 30-35?

File: Graph-Convolutional-Networks-Model-Relations-In-Data/README.md [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# Training
+For training run [jupyter notebook](graph_convolutional_networks_model_relations_in_data.ipynb)
+
+_Note: we use dataset from [this post](../PyTorch-Multi-Label-Image-Classification-Image-Tagging/README.md), so please refer to it for a detailed description and instructions._
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/README.md:4]: the dataset?

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 14]
```diff
@@ -0,0 +1,685 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# Download GloVe model trained on wikipedia.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:119]: Remove this line.

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 15]
```diff
@@ -0,0 +1,685 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Calculate GloVe embeddings for each label in our target label subset.\n",
+    "small_labels = ['house', 'birds', 'sun', 'valley',\n",
+    "               'nighttime', 'boats', 'mountain', 'tree', 'snow', 'beach', 'vehicle', 'rocks',\n",
+    "               'reflection', 'sunset', 'road', 'flowers', 'ocean', 'lake', 'window', 'plants',\n",
+    "               'buildings', 'grass', 'water', 'animal', 'person', 'clouds', 'sky']\n",
+    "vectorised_labels = [embeddings_dict[label].tolist() for label in small_labels]\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:165]: vectorized

File: Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb [hunk 16]
```diff
@@ -0,0 +1,685 @@
...
+    "    plt.show()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Now we can draw t-SNE visualisation.\n",
```
Comment (tkhanova) [inline Graph-Convolutional-Networks-Model-Relations-In-Data/graph_convolutional_networks_model_relations_in_data.ipynb:213]: visualization

File: README.md [hunk 17]
```diff
@@ -13,6 +13,7 @@ Want to become an expert in AI? [AI Courses by OpenCV](https://opencv.org/course
 
 | Blog Post | |
 | ------------- |:-------------|
+|[Graph Convolutional Networks: Model Relations In Data](https://www.learnopencv.com/graph-convolutional-networks:-model-relations-in-data/)|[Code](https://github.com/spmallick/learnopencv/tree/master/Graph-Convolutional-Networks-Model-Relations-In-Data)|
```
Comment (tkhanova) [inline README.md:16]: https://www.learnopencv.com/graph-convolutional-networks-model-relations-in-data/
